src/v/kafka/requests/fetch_request.cc

```diff
@@ -293,31 +294,34 @@ static ss::future<fetch_response::partition_response> read_from_partition(
       std::nullopt);
 
     reader_config.strict_max_bytes = config.strict_max_bytes;
+    auto hw = pw.high_watermark();
+    auto lso = pw.last_stable_offset();
 
     return pw.make_reader(reader_config, deadline)
-      .then([pw, timeout = config.timeout](
-              model::record_batch_reader reader) mutable {
-          vlog(klog.trace, "fetch reader {}", reader);
-          return std::move(reader)
-            .consume(kafka_batch_serializer(), timeout)
-            .then([pw](kafka_batch_serializer::result res) mutable {
```

Comment:
I'm not sure I understand the need for the foreign-aware handling here. Above, on the partition's home core we consume the batches locally into a new, unshared iobuf during serialization then that resulting iobuf is moved back to the core handling the request, right?

i think adding some more detail for this issue to the commit message is necessary otherwise new readers of this code won't be able to deduce the issue.

Reply:
I will add comment to the commit message.
Batches coming from `batch_cache` are 'shared' i.e. we use the same underlying fragments for batch being in the cache and the one that is used to create serialized batch. When we move `iobuf` into serialized batch version we are still using the same fragments that were underlying the original batch still being in the cache. The control structure for `iobufs` being underlie by the same fragments live on two different cores causing race condition when accessing fragments.

Reply:
i see. i had probably thought that moving the iobuf into the serializer changed ownership. that'd be a great paragraph to add to the commit message, please. i think this code then makes sense.

Reply:
ack, I have included that in the commit message